cpp/src/copying/gather_opt.cu

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2018-2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gather.hpp"
+#include <cudf/copying.hpp>
+#include <cudf/cudf.h>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/cudf_utils.h>
+#include <utilities/type_dispatcher.hpp>
+#include <bitmask/legacy_bitmask.hpp>
+#include <cudf/table.hpp>
+#include <string/nvcategory_util.hpp>
+
+#include <algorithm>
+#include <thrust/gather.h>
+
+#include <table/device_table.cuh>
+
+#include <cudf/table.hpp>
+
+#include <utilities/column_utils.hpp>
+
+/**
+ * @brief Operations for copying from one column to another
+ * @file copying_ops.cu
+ */
+
+namespace impl {
+
+template<class T>
+__device__ inline void warp_wise_reduce(T& f) {
+  #pragma unroll
+  for(int offset = 16; offset > 0; offset /= 2){
+    // TODO: ONLY works for CUDA 9.2 or later
+    float other_f = __shfl_down_sync(0xffffffffu, f, offset);
+    f += other_f;
+  }
+}
+
+template<class T>
+__device__ inline void block_wise_reduce(T f, T* smem){
+  __syncthreads();
+  int lane_id = (threadIdx.x & 31);
+  int warp_id = (threadIdx.x >> 5);
+  // Assuming lane 0 of each warp holds the value that we want to perform reduction
+  if(lane_id == 0){
+    smem[warp_id] = f;
+  }
+  __syncthreads();
+  if(warp_id == 0){ // Here I am assuming maximum block size is 1024 and 1024 >> 5 = 32
+    f = (lane_id < (blockDim.x>>5)) ? smem[lane_id] : 0;
+    warp_wise_reduce(f); 
+    if(lane_id == 0){
+      smem[0] = f;
+    }
+  }
+  __syncthreads();
+}
+
+template<bool check_bounds>
+__device__ __inline__ void gather_bitmask_device(
+    gdf_valid_type const* const __restrict__ source_mask,
+    gdf_size_type const source_row,
+    gdf_size_type const num_source_rows, 
+    gdf_valid_type* const destination_mask,
+    gdf_size_type const destination_row, 
+    gdf_size_type const num_destination_rows,
+    gdf_size_type* smem
+  ){
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK = sizeof(MaskType) * 8;
+
+  // Cast bitmask to a type to a 4B type
+  // TODO: Update to use new bit_mask_t
+  MaskType* const __restrict__ destination_mask32 =
+      reinterpret_cast<MaskType*>(destination_mask);
+
+  auto active_threads =
+      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  bool source_bit_is_valid;
+  if(check_bounds){
+    source_bit_is_valid = source_row < num_source_rows && gdf_is_valid(source_mask, source_row);
```

Comment:
Use `bit_mask::is_valid()` instead of `gdf_is_valid`. The latter checks the pointer for null every time, and assumes an 8-bit mask type, which is less efficient.